spec/java/util/OptionalInt.lsl

```diff
@@ -0,0 +1,282 @@
+libsl "1.1.0";
+
+library "std:collections"
+    version "11"
+    language "Java"
+    url "-";
+
+// imports
+
+import "java-common.lsl";
+import "java/lang/_interfaces.lsl";
+import "java/util/function/_interfaces.lsl";
+import "java/util/stream/_interfaces.lsl";
+
+
+// local semantic types
+
+
+
+// automata
+
+@WrapperMeta(
+    src="java.util.OptionalInt",
+    dst="ru.spbpu.libsl.overrides.collections.OptionalInt",
+)
+@public automaton OptionalInt: int
+(
+    var value: int = 0;
+    var present: boolean = false;
+)
+{
+    // states and shifts
+
+    initstate Allocated;
+    state Initialized;
+
+    shift Allocated -> Initialized by [
+        // constructors
+        OptionalInt (),
+        OptionalInt (int),
+
+        // static methods
+        empty,
+        of,
+        ofNullable,
+    ];
+
+    shift Initialized -> self by [
+        // read operations
+        getAsInt,
+        isPresent,
+        isEmpty,
+        ifPresent,
+        ifPresentOrElse,
+        filter,
+        map,
+        flatMap,
+        or,
+        stream,
+        orElse,
+        orElseGet,
+        orElseThrow (),
+        orElseThrow (Supplier),
+        toString,
+        hashCode,
+        equals,
+    ];
+
+
+    // constructors
+
+    @private constructor OptionalInt ()
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    @private constructor OptionalInt (x: int)
+    {
+        action ERROR("Private constructor call");
+    }
+
+
+    // utilities
+
+    @CacheStaticOnce
+    @static proc _makeEmpty (): OptionalInt
+    {
+        // #problem
+        result = new OptionalInt(state=Initialized);
+    }
+
+
+    @AutoInline
+    proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    // static methods
+
+    @static fun empty (): OptionalInt
+    {
+        result = _makeEmpty();
+    }
+
+
+    @static fun of (x: int): OptionalInt
+    {
+        result = new OptionalInt(state=Initialized, value=x, present=true);
+    }
+
+
+    // methods
+
+    fun equals (other: Object): boolean
+    {
+        if (other == self)
+        {
+            result = true;
+        }
+        else
+        {
+            val isSameType = action OBJECT_SAME_TYPE(self, other);
+            if (isSameType)
+            {
+                val otherValue = OptionalInt(other).value;
+                val otherPresent = OptionalInt(other).present;
+                result = self.value == otherValue && self.present == otherPresent;
+            }
+            else
+            {
+                result = false;
+            }
+        }
+    }
+
+
+    fun getAsInt (): int
+    {
+        if (!present)
+            action THROW_NEW("java.util.NoSuchElementException", ["No value present"]);
+
+        result = value;
+    }
+
+
+    fun hashCode (): int
```

Comment:
In "OptionalInt" class body of this method looks in such way:

return isPresent ? Double.hashCode(value) : 0;

Do we need add checking of the "isPresent" field ? Or not ?

Reply:
Fixed